### PR TITLE
app flash: allow apps to write their own flash

### PR DESCRIPTION
**Pull request by bradjc**
Provides userspace with a mechanism to write their own flash space.

### Help

I need some way to ensure that an app is only trying to write its own space (and not the TBF header). Also, I think this doesn't work if the MPU is enabled, even if an app is writing its own space.

### CODE REVIEW

**Comment by ppannuto** [comment]
I think architecturally it may make more sense to explicitly put flash structures into their own sections (i.e. `__attribute__((section(".flash.mything"))`), several thoughts here:

  - putting it in a dedicated section makes it easier to ensure that multiple languages (i.e. future userland rust) can use the same mechanism
  - This makes marking the region of flash relatively trivial, as well as supporting multiple structures instead of just one:
    - The linker can group together all of `.flash.*` into one section
    - The linker can force that section to be on an MPU-friendly alignment
    - The linker can put that section start and length into the ELF, which we can then roll into the TBF, so the kernel can easily configure a R/W MPU region for flash (pretty sure all of flash is currently RO+X, which is why this doesn't work with an MPU -- this will also solve the protecting TBF header problem
  - _technically_, `const` doesn't have any guarantees about where data is placed. In practice, I'd be shocked to see gcc put a global const anywhere except the RO data segment, but there's no requirement that it does

**Comment by ppannuto** [comment]
I also don't really like the current user interface, it's a little weird to me to have this hidden state of stored pointers in the library. Why not just have a library call that has both an `allow` and `subscribe` where you pass the flash and RAM, i.e. 

```c
app_flash_read(const void* flash_ptr, void* ram_ptr);
app_flash_read_async(const void* flash_ptr, void* ram_ptr, callback);
app_flash_write(const void* flash_ptr, const void* ram_ptr);
app_flash_write_async(const void* flash_ptr, const void* ram_ptr, callback)
```

And no `configure` method anymore

**Comment by bradjc** [comment]
> make more sense to explicitly put flash structures into their own sections

That all seems fine, are you willing to add a commit? `const` was a lot easier to implement.

> pretty sure all of flash is currently RO+X

The `write` executes in the kernel, so we could temporarily change the MPU. We need to do something like that for app loading too.

> I also don't really like the current user interface

The interface is intentionally designed to support the expected workflow. It is not designed for arbitrary flash access. My thought was that users would have a struct that they want to be persistent, and they would call `read` at the beginning and periodically call `write` to checkpoint/save it. Saving to app flash is not designed for arbitrary or data storage, and the interface reflects that.

**Comment by bradjc** [comment]
If you want to test this it is helpful to have commit 269c587f6761c9f590928d6db62a7ec83ca4a6c7 to add this to hail.

**Comment by bradjc** [comment]
Without the MPU disabled. "Setting new bytes" is when it calls `write()`.

```
My flash buffer address: 0x31704
Read bytes:
0x9 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 
Setting new bytes.


Kernel panic at /home/bradjc/git/tock/chips/sam4l/src/lib.rs:278:
	"Kernel HardFault.
	Kernel version 269c587f
	r0  0x18b
	r1  0x20001630
	r2  0x31600
	r3  0x20000e80
	r12 0xffffffff
	lr  0xaa000024
	pc  0x15710
	prs 0x1000000 [ N 0 Z 0 C 0 V 0 Q 0 GE 0000 ; ICI.IT 0 T true ; Exc 0-Thread Mode ]
	sp  0x200032b8
	top of stack     0x20003510
	bottom of stack  0x2000250c
	SHCSR 0x0
	CFSR  0x82
	HSFR  0x40000000
	Instruction Access Violation:       false
	Data Access Violation:              true
	Memory Management Unstacking Fault: false
	Memory Management Stacking Fault:   false
	Memory Management Lazy FP Fault:    false
	Instruction Bus Error:              false
	Precise Data Bus Error:             false
	Imprecise Data Bus Error:           false
	Bus Unstacking Fault:               false
	Bus Stacking Fault:                 false
	Bus Lazy FP Fault:                  false
	Undefined Instruction Usage Fault:  false
	Invalid State Usage Fault:          false
	Invalid PC Load Usage Fault:        false
	No Coprocessor Usage Fault:         false
	Unaligned Access Usage Fault:       false
	Divide By Zero:                     false
	Bus Fault on Vector Table Read:     false
	Forced Hard Fault:                  true
	Faulting Memory Address: (valid: true) 0x00031600
	Bus Fault Address:       (valid: false) 0x00031600
"
	Kernel version 269c587f

---| Fault Status |---
No faults detected.

---| App Status |---
App: app_flash   -   [Running]
 Events Queued: 0   Syscall Count: 20   Last Syscall: COMMAND

 ╔═══════════╤══════════════════════════════════════════╗
 ║  Address  │ Region Name    Used | Allocated (bytes)  ║
 ╚0x20006000═╪══════════════════════════════════════════╝
             │ ▼ Grant         372 |   1024          
  0x20005E8C ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ Unused
  0x200051F8 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ ▲ Heap         1808 |   1024 EXCEEDED!    S
  0x20004AE8 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ R
             │ Data            744 |    744              A
  0x20004800 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ M
             │ ▼ Stack         512 |   2048          
  0x20004600 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ Unused
  0x20004000 ┴───────────────────────────────────────────
             .....
  0x00032000 ┬───────────────────────────────────────────
             │ Unused
  0x00031914 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ F
             │ Data            281                       L
  0x000317FB ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ A
             │ Text           6023                       S
  0x00030074 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ H
             │ Header          116
  0x00030000 ┴───────────────────────────────────────────

  R0 : 0x00000000    R6 : 0x00000000
  R1 : 0x00000001    R7 : 0x00000000
  R2 : 0x00031704    R8 : 0x00000000
  R3 : 0x20004A58    R10: 0x00000000
  R4 : 0x20004A7C    R11: 0x00000000
  R5 : 0x00000041    R12: 0x00000000
  R9 : 0x20004800 (Static Base Register)
  SP : 0x200047C0 (Process Stack Pointer)
  LR : 0x0003010F [0x8000009A in lst file]
  PC : 0x000301C6 [0x80000152 in lst file]
 YPC : 0x000301A0 [0x8002EA18 in lst file]

 APSR: N 0 Z 0 C 1 V 0 Q 0
       GE 0 0 0 0
 IPSR: Exception Type - Thread Mode
 EPSR: ICI.IT 0x00
       ThumbBit true 
```

Interestingly, afterwards when I reset the board I get an immediate error:

```
Kernel panic at /home/bradjc/git/tock/kernel/src/process.rs:317:
	"Process app_flash had a fault"
	Kernel version 269c587f

---| Fault Status |---
Data Access Violation:              true
Forced Hard Fault:                  true
Faulting Memory Address:            0x80030073
Fault Status Register (CFSR):       0x00000082
Hard Fault Status Register (HFSR):  0x40000000

---| App Status |---
App: app_flash   -   [Fault]
 Events Queued: 0   Syscall Count: 3   Last Syscall: MEMOP

 ╔═══════════╤══════════════════════════════════════════╗
 ║  Address  │ Region Name    Used | Allocated (bytes)  ║
 ╚0x20006000═╪══════════════════════════════════════════╝
             │ ▼ Grant         272 |   1024          
  0x20005EF0 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ Unused
  0x200050A4 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ ▲ Heap         1468 |   1024 EXCEEDED!    S
  0x20004AE8 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ R
             │ Data            744 |    744              A
  0x20004800 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ M
             │ ▼ Stack         368 |   2048          
  0x20004690 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
             │ Unused
  0x20004000 ┴───────────────────────────────────────────
             .....
  0x00032000 ┬───────────────────────────────────────────
             │ Unused
  0x00031914 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ F
             │ Data            281                       L
  0x000317FB ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ A
             │ Text           6023                       S
  0x00030074 ┼┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ H
             │ Header          116
  0x00030000 ┴───────────────────────────────────────────

  R0 : 0x00000000    R6 : 0x80030073
  R1 : 0x00000001    R7 : 0x00000000
  R2 : 0x00000030    R8 : 0x00000000
  R3 : 0x20004740    R10: 0x00000000
  R4 : 0x80030073    R11: 0x00000000
  R5 : 0x20004800    R12: 0x00000000
  R9 : 0x20004800 (Static Base Register)
  SP : 0x20004700 (Process Stack Pointer)
  LR : 0x00031075 [0x80001000 in lst file]
  PC : 0x00030CD0 [0x80000C5C in lst file]
 YPC : 0x00030075 [0x8002E8EE in lst file]

 APSR: N 1 Z 0 C 0 V 0 Q 0
       GE 0 0 0 0
 IPSR: Exception Type - Thread Mode
 EPSR: ICI.IT 0x00
       ThumbBit true
```

**Comment by ppannuto** [comment]
@bradjc I rebased this onto latest master and added a commit that lays the groundwork for putting application state into a dedicated linker section (need to update the linker script and elf2tbf still)

**Comment by bradjc** [comment]
OK this is update. It is currently rebased onto #396 because it needs that to run, but once that is merged this is only 4 commits.

To test you want 88a1609bb55b5c1b37db2a7e411e8446fe8bb387 for hail.

This defines a new section in the linker (`app_state`) and updates elf2tbf to add that to the TBF header. Userspace then gets a nice interface to be able to make a struct persistent.

# Help!

This PR needs a way to turn off or somehow change the MPU when in kernel mode. While we set the `PRIVDEFENA` bit in the control register, that just disables the MPU for any address space that isn't covered by an active region. The kernel cannot actually write the flash space of an app.

**Comment by ppannuto** [comment]
[This isn't an answer, this is just a collection of information]

I'm a little confused at the behavior you're seeing. If I understand correctly, you have things working if you disable the initial call to `enable_mpu`? That doesn't totally jive with my understanding of the MPU. 

From the docs:
image

We're setting `PRIVDEFENA` to 1:

```rust
        // Enable the MPU, disable it during HardFault/NMI handlers, disable it
        // when privileged code runs
        regs.control.set(0b101);
```

Which you noted, but apparently isn't working. However, if you disable the MPU:
image

It should be using the exact same memory map as the `PRIVDEFENA=1` case.

Following the documentation for the default memory map, we find these two important bits:

 - ARM specifies a few regions of memory as XN (execute never), but ARM specifies nothing about the RW state of these memory regions:

image

 - ARM does not specify anything about the R/W state of the default memory map:
image

But I couldn't find anything in the SAM4L docs about their definition, so I'm assuming that they just leave the whole address space as RW by default.

-------------

Questions:

1. Does this work correctly if you never call `enable_mpu`?

2. If you add a `disable_mpu` and call that before writing flash, does that make things work?

```rust
    fn disable_mpu(&self) {
        let regs = unsafe { &*self.0 };
        regs.control.set(0b0);
    }
```

**Comment by bradjc** [comment]
1. Yes.
2. Yes.

    I made the following diff:

    ```diff
	diff --git a/kernel/src/sched.rs b/kernel/src/sched.rs
	index a871fb52..905be28c 100644
	--- a/kernel/src/sched.rs
	+++ b/kernel/src/sched.rs
	@@ -8,6 +8,7 @@ use process;
	 use process::{Process, Task};
	 use returncode::ReturnCode;
	 use syscall::Syscall;
	+use platform::mpu::MPU;

	 pub unsafe fn do_process<P: Platform, C: Chip>(platform: &P,
	                                                chip: &mut C,
	@@ -32,6 +33,8 @@ pub unsafe fn do_process<P: Platform, C: Chip>(platform: &P,
	                 systick.enable(false);
	             }
	             process::State::Yielded => {
	+                chip.mpu().disable_mpu();
	+
	                 match process.dequeue_task() {
	                     None => break,
	                     Some(cb) => {
    ```
    And the app works.

I think the comment in our source is bogus, and that the key line is:

> All memory regions configured in the MPU take priority over the default memory map.

So setting the `PRIVDEFENA` bit is very different from disabling the MPU.

Is there a safe place to call `mpu.disable()`?

**Comment by ppannuto** [comment]
Can it simply be disabled at the beginning of the flash write and re-enabled at the end of the write? Or is there some DMA/other trickiness going on?

**Comment by bradjc** [comment]
This is updated (use commit 81e32324 to use this with hail).

Not much has changed, but it is rebased on master which has some new commits that are necessary for this to work.

File: kernel/src/process.rs [hunk 1]
```diff
@@ -100,6 +100,27 @@ pub fn schedule(callback: FunctionCall, appid: AppId) -> bool {
...
+pub fn get_editable_flash_range(app_idx: usize) -> (usize, usize) {
+    let procs = unsafe { &mut PROCS };
+    if app_idx >= procs.len() {
+        return (0, 0);
+    }
+
+    match procs[app_idx] {
+        None => (0, 0),
+        Some(ref mut p) => {
+            // TODO(alevy): validate appid liveness
```
Comment (brghena) [inline kernel/src/process.rs]: What's this?

File: userland/libtock/app_state.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+//
+//     APP_STATE_DECLARE(struct my_state_t, memory_copy);
+//
+// Which you can read as the equivalent of:
+//
+//     struct my_state_t memory_copy;
+//
+// The variable `memory_copy` is available as regular C structure, however
+// users must explicitly `load` and `save` application state as appropriate
+#define APP_STATE_DECLARE(_type, _identifier)                         \
```
Comment (brghena) [inline userland/libtock/app_state.h]: You should note that users can only call APP_STATE_DECLARE once per process.

Comment (brghena) [inline userland/libtock/app_state.h]: Taken care of

File: kernel/src/process.rs (hunk 1 above)
Comment (bradjc) [inline kernel/src/process.rs]: Existing comment that I copied. Issue #526.

**Pull request merged**